File: results/dnd-rest-benchmark/lifecycle-runs/20260731T062307Z_claude_claude-sonnet-5_typescript-node/src/routes/quests.ts
// Campaign quest tracking: create quests, record milestone progress, and
// summarize quest counts by status. Persistent (`campaign_quests`).
import type { ServerResponse } from "node:http";
import { db } from "../db.js";
import { sendJson } from "../http.js";
import { isPlainObject } from "../validation.js";
import { hasCampaign } from "./campaigns.js";

const VALID_STATUSES = new Set(["active", "completed", "blocked"]);

interface QuestRecord {
  campaignId: string;
  id: string;
  title: string;
  status: string;
  milestones: string[];
  milestonesDone: string[];
}

function getQuest(campaignId: string, id: string): QuestRecord | undefined {
  const row = db
    .prepare("SELECT campaign_id, id, title, status, milestones, milestones_done FROM campaign_quests WHERE campaign_id = ? AND id = ?")
    .get(campaignId, id) as
    | { campaign_id: string; id: string; title: string; status: string; milestones: string; milestones_done: string }
    | undefined;
  if (!row) return undefined;
  return {
    campaignId: row.campaign_id,
    id: row.id,
    title: row.title,
    status: row.status,
    milestones: JSON.parse(row.milestones) as string[],
    milestonesDone: JSON.parse(row.milestones_done) as string[],
  };
}

function saveQuest(quest: QuestRecord): void {
  db.prepare(
    "INSERT INTO campaign_quests (campaign_id, id, title, status, milestones, milestones_done) VALUES (?, ?, ?, ?, ?, ?)",
  ).run(
    quest.campaignId,
    quest.id,
    quest.title,
    quest.status,
    JSON.stringify(quest.milestones),
    JSON.stringify(quest.milestonesDone),
  );
}

function updateQuestProgress(quest: QuestRecord): void {
  db.prepare("UPDATE campaign_quests SET milestones_done = ? WHERE campaign_id = ? AND id = ?").run(
    JSON.stringify(quest.milestonesDone),
    quest.campaignId,
    quest.id,
  );
}

function getCampaignQuests(campaignId: string): QuestRecord[] {
  const rows = db
    .prepare("SELECT campaign_id, id, title, status, milestones, milestones_done FROM campaign_quests WHERE campaign_id = ?")
    .all(campaignId) as { campaign_id: string; id: string; title: string; status: string; milestones: string; milestones_done: string }[];
  return rows.map((row) => ({
    campaignId: row.campaign_id,
    id: row.id,
    title: row.title,
    status: row.status,
    milestones: JSON.parse(row.milestones) as string[],
    milestonesDone: JSON.parse(row.milestones_done) as string[],
  }));
}

export function handleCreateQuest(res: ServerResponse, campaignId: string, body: unknown): void {
  if (!hasCampaign(campaignId)) {
    sendJson(res, 404, { error: "campaign not found" });
    return;
  }

  if (
    !isPlainObject(body) ||
    typeof body.id !== "string" ||
    !body.id ||
    typeof body.title !== "string" ||
    typeof body.status !== "string" ||
    !VALID_STATUSES.has(body.status) ||
    !Array.isArray(body.milestones) ||
    !body.milestones.every((m) => typeof m === "string")
  ) {
    sendJson(res, 400, { error: "invalid request" });
    return;
  }

  if (getQuest(campaignId, body.id)) {
    sendJson(res, 409, { error: "quest already exists" });
    return;
  }

  const quest: QuestRecord = {
    campaignId,
    id: body.id,
    title: body.title,
    status: body.status,
    milestones: body.milestones as string[],
    milestonesDone: [],
  };
  saveQuest(quest);

  sendJson(res, 201, {
    id: quest.id,
    title: quest.title,
    status: quest.status,
    milestones_total: quest.milestones.length,
    milestones_done: quest.milestonesDone.length,
  });
}

export function handleUpdateQuestProgress(
  res: ServerResponse,
  campaignId: string,
  questId: string,
  body: unknown,
): void {
  if (!hasCampaign(campaignId)) {
    sendJson(res, 404, { error: "campaign not found" });
    return;
  }

  const quest = getQuest(campaignId, questId);
  if (!quest) {
    sendJson(res, 404, { error: "quest not found" });
    return;
  }

  if (
    !isPlainObject(body) ||
    !Array.isArray(body.completed) ||
    !body.completed.every((m) => typeof m === "string") ||
    !(body.completed as string[]).every((m) => quest.milestones.includes(m))
  ) {
    sendJson(res, 400, { error: "invalid request" });
    return;
  }

  const done = new Set(quest.milestonesDone);
  for (const milestone of body.completed as string[]) {
    done.add(milestone);
  }
  quest.milestonesDone = quest.milestones.filter((m) => done.has(m));
  updateQuestProgress(quest);

  sendJson(res, 200, {
    id: quest.id,
    status: quest.status,
    milestones_total: quest.milestones.length,
    milestones_done: quest.milestonesDone.length,
  });
}

export function handleQuestSummary(res: ServerResponse, campaignId: string): void {
  if (!hasCampaign(campaignId)) {
    sendJson(res, 404, { error: "campaign not found" });
    return;
  }

  const quests = getCampaignQuests(campaignId);
  const summary = { active: 0, completed: 0, blocked: 0 };
  for (const quest of quests) {
    if (quest.status === "active") summary.active += 1;
    else if (quest.status === "completed") summary.completed += 1;
    else if (quest.status === "blocked") summary.blocked += 1;
  }

  sendJson(res, 200, {
    campaign_id: campaignId,
    active: summary.active,
    completed: summary.completed,
    blocked: summary.blocked,
  });
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T052954Z_claude_claude-sonnet-5_python-flask/dndapp/routes/quests.py
"""Campaign quest tracking: create quests, record milestone progress, summarize."""

import json

from flask import Blueprint, jsonify, request

from ..db import get_db
from ..validation import valid_nonempty_str

bp = Blueprint("quests", __name__)

VALID_STATUSES = ("active", "completed", "blocked")


def _valid_milestones(value):
    return (
        isinstance(value, list)
        and len(value) > 0
        and all(valid_nonempty_str(m) for m in value)
    )


@bp.post("/v1/campaigns/<campaign_id>/quests")
def create_quest(campaign_id):
    conn = get_db()
    try:
        campaign = conn.execute(
            "SELECT id FROM campaigns WHERE id = ?", (campaign_id,)
        ).fetchone()
        if campaign is None:
            return jsonify(error="campaign not found"), 404

        data = request.get_json(silent=True) or {}
        quest_id = data.get("id")
        title = data.get("title")
        status = data.get("status")
        milestones = data.get("milestones")

        if not isinstance(quest_id, str) or not quest_id:
            return jsonify(error="invalid id"), 400
        if not valid_nonempty_str(title):
            return jsonify(error="invalid title"), 400
        if status not in VALID_STATUSES:
            return jsonify(error="invalid status"), 400
        if not _valid_milestones(milestones):
            return jsonify(error="invalid milestones"), 400

        existing = conn.execute(
            "SELECT id FROM quests WHERE campaign_id = ? AND id = ?",
            (campaign_id, quest_id),
        ).fetchone()
        if existing is not None:
            return jsonify(error="quest already exists"), 409

        conn.execute(
            "INSERT INTO quests (campaign_id, id, title, status, milestones_json, completed_json) "
            "VALUES (?, ?, ?, ?, ?, ?)",
            (campaign_id, quest_id, title, status, json.dumps(milestones), json.dumps([])),
        )
        conn.commit()
    finally:
        conn.close()

    return jsonify(
        id=quest_id,
        title=title,
        status=status,
        milestones_total=len(milestones),
        milestones_done=0,
    ), 201


@bp.post("/v1/campaigns/<campaign_id>/quests/<quest_id>/progress")
def update_quest_progress(campaign_id, quest_id):
    conn = get_db()
    try:
        quest = conn.execute(
            "SELECT * FROM quests WHERE campaign_id = ? AND id = ?",
            (campaign_id, quest_id),
        ).fetchone()
        if quest is None:
            return jsonify(error="quest not found"), 404

        data = request.get_json(silent=True) or {}
        completed = data.get("completed")

        if not isinstance(completed, list) or not all(
            isinstance(m, str) for m in completed
        ):
            return jsonify(error="invalid completed"), 400

        milestones = json.loads(quest["milestones_json"])
        already_done = json.loads(quest["completed_json"])

        newly_done = set(already_done)
        for milestone in completed:
            if milestone in milestones:
                newly_done.add(milestone)

        done_list = [m for m in milestones if m in newly_done]

        conn.execute(
            "UPDATE quests SET completed_json = ? WHERE campaign_id = ? AND id = ?",
            (json.dumps(done_list), campaign_id, quest_id),
        )
        conn.commit()
    finally:
        conn.close()

    return jsonify(
        id=quest_id,
        status=quest["status"],
        milestones_total=len(milestones),
        milestones_done=len(done_list),
    )


@bp.get("/v1/campaigns/<campaign_id>/quests/summary")
def quest_summary(campaign_id):
    conn = get_db()
    try:
        campaign = conn.execute(
            "SELECT id FROM campaigns WHERE id = ?", (campaign_id,)
        ).fetchone()
        if campaign is None:
            return jsonify(error="campaign not found"), 404

        rows = conn.execute(
            "SELECT status FROM quests WHERE campaign_id = ?", (campaign_id,)
        ).fetchall()
    finally:
        conn.close()

    counts = {"active": 0, "completed": 0, "blocked": 0}
    for row in rows:
        if row["status"] in counts:
            counts[row["status"]] += 1

    return jsonify(
        campaign_id=campaign_id,
        active=counts["active"],
        completed=counts["completed"],
        blocked=counts["blocked"],
    )
